webgrid/filters.py

```diff
@@ -547,67 +631,117 @@ def description(self):
-        return _('{descriptor}{date}',
-                 descriptor=prefix,
-                 date=first_day.strftime('%m/%d/%Y'))
-
-
-class DateFilter(FilterBase, _DateMixin):
+
+class DateOpQueryMixin:
+    def today_query(self, query):
+        today = self._get_today()
+        return query.filter(self.sa_col == today)
+
+    def thisweek_query(self, query):
+        today = self._get_today()
+        return query.filter(self.sa_col.between(
+            today - relativedelta(weekday=SU(-1)),
+            today + relativedelta(weekday=calendar.SATURDAY),
+        ))
+
+    def selmonth_query(self, query):
+        if self.value1 and self.value2:
+            return self._month_year_filter(query)
+        else:
+            return query
+
+    def thismonth_query(self, query):
+        return self._month_year_filter(query)
+
+    def lastmonth_query(self, query):
+        return self._month_year_filter(query)
+
+    def thisyear_query(self, query):
+        return self._month_year_filter(query)
+
+    def between_query(self, query):
+        return query.filter(self._between_clause())
+
+    def notbetween_query(self, query):
+        return query.filter(~self._between_clause())
+
+    def da_query(self, query):
+        today = self._get_today()
+        return query.filter(
+            self.sa_col == today - dt.timedelta(days=self.value1)
+        )
+
+    def ltda_query(self, query):
+        today = self._get_today()
+        return query.filter(and_(
+            self.sa_col > today - dt.timedelta(days=self.value1),
+            self.sa_col < today,
+        ))
+
+    def mtda_query(self, query):
+        today = self._get_today()
+        return query.filter(
+            self.sa_col < today - dt.timedelta(days=self.value1)
+        )
+
+    def ind_query(self, query):
+        today = self._get_today()
+        return query.filter(
+            self.sa_col == today + dt.timedelta(days=self.value1)
+        )
+
+    def iltd_query(self, query):
+        today = self._get_today()
+        return query.filter(and_(
+            self.sa_col >= today,
+            self.sa_col < today + dt.timedelta(days=self.value1),
+        ))
+
+    def imtd_query(self, query):
+        today = self._get_today()
+        return query.filter(
+            self.sa_col > today + dt.timedelta(days=self.value1)
+        )
+
+    def _get_today(self):
+        # this filter is date-only, so our "now" is a date without time
+        return ensure_date(self._get_now())
+
+    def _get_now(self):
+        return self._now or dt.datetime.now()
+
+    def _month_year_filter(self, query):
+        return query.filter(self.sa_col.between(self.first_day, self.last_day))
+
+    def _between_clause(self):
+        if self.value1 <= self.value2:
+            return self.sa_col.between(self.value1, self.value2)
+        else:
+            return self.sa_col.between(self.value2, self.value1)
+
+
+class DateFilter(DateOpQueryMixin, _DateMixin, FilterBase):
```

Comment:
We've got two different paradigms happening here:
- dict used as a "switch" block returning lambda, with the operator itself being the key
- mixin of query-modifying methods that get called via method-naming magic

The dict paradigm is, I believe, more sustainable overall and could be used in both places, though I would want to see an ImmutableDict for the class-level definition. Method discovery can work well in some places, but if we were to do it that way, I'd rather have it segmented into its own object and use it through composition rather than inheritance. Here, though, I'd rather see consistency with the chosen paradigm.

Also, with `_DateMixin` as established naming, `DateOpQueryMixin` -> `_DateOpQueryMixin`.

Reply:
I was back and forth on that decision...